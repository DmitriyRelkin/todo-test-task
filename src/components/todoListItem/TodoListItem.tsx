import React from 'react';
import { Text, TouchableOpacity } from 'react-native';
import { ITodo } from "../../store/slices/todoSlice";
import { getTodoItemStyles } from "./styles";

interface ITodoListItemProps {
  todo: ITodo;
  onDeleteTodo: (id: number) => void;
  onSelect: (id: number) => void;
  isSelected: boolean;
}

const TodoListItem: React.FC<ITodoListItemProps> = ({ todo, onSelect, onDeleteTodo, isSelected }) => {
  const { container, title, deleteButton, deleteButtonText, selectedItemStyle } = getTodoItemStyles();

  const handleDeleteTodo = () => {
    onDeleteTodo(todo.id);
  };

  const onPress = () => {
    onSelect(todo.id);
  }

  return (
    <TouchableOpacity testID="todo-item" onPress={onPress} style={[container, isSelected && selectedItemStyle]}>
      <Text style={title}>{todo.text}</Text>
      <TouchableOpacity testID="delete-button" style={deleteButton} onPress={handleDeleteTodo}>
        <Text style={deleteButtonText}>DELETE</Text>
      </TouchableOpacity>
    </TouchableOpacity>
  );
};

export default TodoListItem;
