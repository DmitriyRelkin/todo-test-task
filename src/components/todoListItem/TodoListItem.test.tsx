import React from 'react';
import { TouchableOpacity } from 'react-native';
import renderer from 'react-test-renderer';
import TodoListItem from './TodoListItem';

describe('TodoListItem', () => {
  const todo = {
    id: 1,
    text: 'Buy groceries',
  };
  const onSelect = jest.fn();
  const onDeleteTodo = jest.fn();
  const isSelected = false;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the todo item correctly', () => {
    const tree = renderer.create(
      <TodoListItem todo={todo} onSelect={onSelect} onDeleteTodo={onDeleteTodo} isSelected={isSelected} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('calls onSelect when the item is pressed', () => {
    const component = renderer.create(
      <TodoListItem todo={todo} onSelect={onSelect} onDeleteTodo={onDeleteTodo} isSelected={isSelected} />
    );

    const touchableElement = component.root.findByType(TouchableOpacity);
    touchableElement.props.onPress();

    expect(onSelect).toHaveBeenCalledWith(1);
  });

  it('calls onSelect when the todo item is pressed', () => {
    const component = renderer.create(
      <TodoListItem todo={todo} onSelect={onSelect} onDeleteTodo={onDeleteTodo} isSelected={isSelected} />
    );

    const todoElement = component.root.findByProps({ testID: 'todo-item' });
    todoElement.props.onPress();

    expect(onSelect).toHaveBeenCalledWith(1);
  });

  it('calls onDeleteTodo when the delete button is pressed', () => {
    const component = renderer.create(
      <TodoListItem todo={todo} onSelect={onSelect} onDeleteTodo={onDeleteTodo} isSelected={isSelected} />
    );

    const deleteButtonElement = component.root.findByProps({ testID: 'delete-button' });
    deleteButtonElement.props.onPress();

    expect(onDeleteTodo).toHaveBeenCalledWith(1);
  });
});
