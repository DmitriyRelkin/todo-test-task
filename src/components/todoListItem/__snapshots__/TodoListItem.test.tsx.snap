// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem renders the todo item correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 8,
      "opacity": 1,
      "padding": 10,
      "shadowColor": "#171717",
      "shadowOffset": {
        "height": 4,
        "width": -2,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 3,
    }
  }
  testID="todo-item"
>
  <Text
    style={
      {
        "fontSize": 16,
      }
    }
  >
    Buy groceries
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f89f9f",
        "borderRadius": 12,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 100,
      }
    }
    testID="delete-button"
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontWeight": "bold",
        }
      }
    >
      DELETE
    </Text>
  </View>
</View>
`;
